### ISSUE TITLE
How does resolution scaling work?

### ISSUE THREAD

**Issue by foemre** [ultralytics/yolov5#10546]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have read EfficientDet paper and found that it is possible to scale my input image size during training. But the paper does not explain how resolution scaling works during inference. 

My question is: how is it possible that I can train a model with, say 480x480 input images and perform inference with 640x640 images? Is the model scaled up/down to accomodate larger/smaller images during inference, or is there another mechanism?

**Comment by iceisfun** [comment]
Are you looking for this?

https://github.com/ultralytics/yolov5/blob/96a71b17a276fa0a0b6fbdf68d579ce0603bfa2f/utils/augmentations.py#L111

**Comment by foemre** [comment]
@iceisfun I don't think so, what I mean is: with traditional image classifiers, e.g. with VGG19 the input shape has to be 224x224. If the input image has a different shape, it is rescaled with or without padding, because the first layer accepts tensors of size 224x224. Similarly, if I export my Pytorch model with .pt extension as ONNX/TRT, I have to use the image size I specified. 

But **that is not the case** if I directly use the .pt extension model. I can train with 512x512 images and perform inference with whatever image size I want. 

I think it is related to the resolution scaling / compound scaling as described in EfficientDet paper, I remember @glenn-jocher mentioning the paper in some other issue. But I don't understand how it works or relates to YOLOv5.

I guess I'm trying to ask: how can the input layer's shape change after training?

**Comment by mikel-brostrom** [comment]
> perform inference with whatever image size I want

It has to be a multiple of 32, as this is the maximum stride of the architecture @foemre

**Comment by mikel-brostrom** [comment]
For example, using `--imgsz 417`

```python
python3 detect.py --imgsz 417
detect: weights=yolov5s.pt, source=data/images, data=data/coco128.yaml, imgsz=[417, 417], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 v7.0-70-g589edc7b Python-3.8.15 torch-1.11.0+cu102 CUDA:0 (Quadro P2000, 4032MiB)

Fusing layers... 
Model summary: 224 layers, 7266973 parameters, 0 gradients
WARNING ⚠️ --img-size [417, 417] must be multiple of max stride 32, updating to [448, 448]
image 1/2 /home/mikel.brostrom/yolov5/data/images/bus.jpg: 448x352 4 persons, 1 bus, 10.5ms
image 2/2 /home/mikel.brostrom/yolov5/data/images/zidane.jpg: 256x448 3 persons, 1 tie, 8.2ms
Speed: 0.3ms pre-process, 9.4ms inference, 6.3ms NMS per image at shape (1, 3, 448, 448)
```

You see that the images are reshaped to the closest upper multiple for 32 during inference. On the other hand, if your `imgsz` is a multiple of 32:

```python
python3 detect.py --imgsz 416
detect: weights=yolov5s.pt, source=data/images, data=data/coco128.yaml, imgsz=[416, 416], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 v7.0-70-g589edc7b Python-3.8.15 torch-1.11.0+cu102 CUDA:0 (Quadro P2000, 4032MiB)

Fusing layers... 
Model summary: 224 layers, 7266973 parameters, 0 gradients
image 1/2 /home/mikel.brostrom/yolov5/data/images/bus.jpg: 416x320 4 persons, 1 bus, 10.0ms
image 2/2 /home/mikel.brostrom/yolov5/data/images/zidane.jpg: 256x416 3 persons, 2 ties, 8.0ms
Speed: 0.3ms pre-process, 9.0ms inference, 1.1ms NMS per image at shape (1, 3, 416, 416)
```

It keeps that dimension, always maintaining the aspect ratio

**Comment by foemre** [comment]
@mikel-brostrom yeah sure, because stride is 32 by default. I know this is a weird question, so let me put here what I know:

- Stride is 32 so in both training and inference the image size has to be a multiple of 32
- Image is padded both in training and inference (if mosaic is not selected in hyperparameters), maintaining the aspect ratio
- Model scaling in model cfg scales the depth and width of layers

Then how is it possible that I can train with size 512 (multiple of 32) and perform inference with size 640 (multiple of 32)?

**Comment by mikel-brostrom** [comment]
In [VGG](https://pytorch.org/vision/main/_modules/torchvision/models/vgg.html#vgg11), the input to the last fully connected layer is hardcoded to 4096 @foemre . This means that your input has to be of a specific size so that this constraint in the architecture is met.

Yolov5 is a fully convolutional network. Each layer adapts to the output of the previous one so there are no limitations on the input size more than what you RAM can fit. For this architecture there is no such limitation.

**Comment by foemre** [comment]
@mikel-brostrom Having done research on the topics you mentioned in your comment, I see that the keyword I was missing was FCN. Now I completely understand it. Thanks!
